src/hazelcore/parse/parse.mly

```diff
@@ -0,0 +1,223 @@
+%{
+  let mk_seq operand =
+    Seq.mk operand []
+
+  let mk_binop l op r =
+    let seq = Seq.seq_op_seq l op r in
+    seq
+
+  let mk_typ_ann pat typ =
+    let typ = UHTyp.mk_OpSeq typ in
+    let pat = Seq.nth_operand 0 pat in
+    mk_seq (UHPat.TypeAnn(ErrStatus.NotInHole, pat, typ))
+
+  let mk_pat_parenthesized e =
+    let e = UHPat.mk_OpSeq e in
+    UHPat.Parenthesized(e)
+
+  let mk_application e args =
+    let e = mk_seq e in
+    let rec mk_app e args =
+      match args with
+      | [] -> e
+      | x::xs -> (
+        let x = mk_seq x in
+        let opseq = mk_app x xs in
+        mk_binop e Operators_Exp.Space opseq
+      )
+    in
+    mk_app e args
+
+  let mk_let_line pat expr =
+    let pat = UHPat.mk_OpSeq pat in
+    UHExp.letline pat expr
+
+  let mk_typ_paren typ =
+    let opseq = UHTyp.mk_OpSeq typ in
+    UHTyp.Parenthesized opseq
+
+  let mk_typ_list typ =
+    let opseq = UHTyp.mk_OpSeq typ in
+    UHTyp.List opseq
+
+  let mk_inj_l expr =
+    UHExp.Inj(ErrStatus.NotInHole, InjSide.L, expr)
+
+  let mk_inj_r expr =
+    UHExp.Inj(ErrStatus.NotInHole, InjSide.R, expr)
+
+  let mk_fn pat expr =
+    let pat = UHPat.mk_OpSeq pat in
+    UHExp.lam pat expr
+
+  let mk_rule pat expr =
+    let pat = UHPat.mk_OpSeq pat in
+    UHExp.Rule(pat, expr)
+
+  let mk_case expr rules =
+    let e = UHExp.case expr rules in
+    mk_seq e
+
+  let mk_empty_list =
+    mk_seq (UHExp.listnil ())
+%}
+
+%token LET
+%token IN
+%token <string> INT
+%token <string> FLOAT
+%token TRUE FALSE
+%token PLUS MINUS
+%token MULT DIV
+%token FPLUS FMINUS
+%token FMULT FDIV
+%token COLON
+%token COLONCOLON
+%token SEMICOLON
+%token EQUAL
+%token GREATER LESSER
+%token PERIOD
+%token COMMA
+%token INJL INJR
+%token EOF
+%token <string> IDENT
+%token LPAREN RPAREN
+%token LBRACE RBRACE
+%token LBRACK RBRACK
+%token EMPTY_HOLE
+%token LAMBDA
+%token CASE
+%token BAR
+%token ARROW
+%token TARROW
+%token END
+%token <string> COMMENT
+%token EMPTY
+
+%left LESSER GREATER EQUAL
+%left PLUS MINUS FPLUS FMINUS
+%left MULT DIV FMULT FDIV
+%right COLONCOLON
+%left BAR
+%right TARROW
+%left COMMA
+
+%start main
+%type <UHExp.t> main
+%%
+
+main:
+  line EOF { $1 }
+  | line SEMICOLON main EOF { List.concat [$1; $3] }
+;
+
+let_binding:
+  LET pat EQUAL line IN { mk_let_line $2 $4 }
+  | LET pat COLON typ EQUAL line IN { mk_let_line (mk_typ_ann $2 $4) $6 }
+;
+
+typ:
+  typ typ_op typ { mk_binop $1 $2 $3 }
+  | typ_ { mk_seq $1 }
+;
+
+typ_:
+  atomic_type { $1 }
+  | LPAREN typ RPAREN { mk_typ_paren $2 }
+  | LBRACK typ RBRACK { mk_typ_list $2 }
+;
+
+atomic_type:
+  IDENT {
+    match $1 with
+    | "Int" -> UHTyp.Int
+    | "Bool" -> UHTyp.Bool
+    | "Float" -> UHTyp.Float
+    | _ -> failwith ("Unknown Type: "^$1)
+  }
+  | EMPTY_HOLE { UHTyp.Hole }
+;
+
+%inline typ_op:
+  COMMA { Operators_Typ.Prod }
+  | TARROW { Operators_Typ.Arrow }
+  | BAR { Operators_Typ.Sum }
+;
+
+pat:
+  pat COLONCOLON pat { mk_binop $1 Operators_Pat.Cons $3 }
```

Comment:
type annotations can now appear on any pattern, not just at the top level of a let or lambda